# SC2 Mutation Frequency Calculator (SMFC)
**Lineage-specific SARS-CoV-2 mutation frequency analysis**
The **SARS-CoV-2 Mutation Frequency Calculator (SC2MFC)** computes dynamic mutation frequencies for Pango lineages using lineage mutation profiles from **covsonar** or other DESH exports.
SC2MFC can be utilised in genomic surveillance pipelines, wastewater analyses, primer design workflows, and lineage characterization efforts. 

---

## Key Features

- Generate mutation frequency matrices (AA or DNA level)
- Export formats: **XLSX**, **CSV**, **TSV**
- Optional Excel gradient heatmaps
- Automatic grouping of rare lineages into **other lineages**
- Frequency and sample size cutoffs (default: 75% & 10 genomes)
- Consensus sequence reconstruction
- Additional analytics:  
  - Sample-size abundance plot  
  - Mutation prevalence plot  
  - Lab/country diversity  
  - Frequency-cutoff validation (precision/recall)

---

## Installation & Setup

### Prerequisites
Conda and python are necessary.
[CovSonar](https://github.com/rki-mf1/covsonar) is the key tool to extract mutation profiles from. It has a database with consensussequences sequenced in the course of the pandemic assigned to lineages which can be queried. 

### Install
Proceed as follows to install SMFC:
```bash
# download the repository to the current working directory using git 
git clone https://github.com/rki-mf1/sc2-mutation-frequency-calculator.git
# build the custom software environment using conda [recommended]
conda env create -n sc2mfc -f sc2-mutation-frequency-calculator/envs/sc2mfc.yml
# activate the conda evironment if built 
conda activate sc2mfc
```

### Options

Show all options:
```bash
python3 scripts/init.py --help
```

| flag                         | description                                                           |   
|------------------------------|-----------------------------------------------------------------------| 
| --mutation_profile           | mutation_profile (TSV) generated by covsonar                          |   
| --matrix                     | create mutation frequency matrix with lineage from defined timeframe  |     
| --matrix_format              | output format for the mutation frequency matrix (default: xlsx)       | 
| --gradient                   | 2-color conditional formatting (only for xlsx)                        | 
| --aa_mutations               | choose specific aa mutations which are set in TXT file                |      
| --cut_off_frequency          | threshold for aa mutations to create matrix or consensus              |     
| --cut_off_lineage            | threshold for number of samples sequenced from a lineage              |     
| --consensus                  | create consensus sequences based on given lineages                    |          
| --signature                  | for given (set of) lineage return the set of signature mutations      |     
| --verify_frequency_cut_off   | evaluate performance of the frequency cut-off by precision and recall |    
| --sample_size_abundancy_plot | plot the sample abundancies of all lineages within the timeframe      | 
 
---

## Example: Generate a Mutation Frequency Matrix 
Typical input is a covSonar mutation profile (TSV). Exmaple input files are stored in input/. 

How a frequency matrix can be created: 
```bash
python3 scripts/init.py \
  -tsv input/covsonar-rki-2023-02-01--2023-04-01.tsv \
  -m -level aa \
  --matrix_format xlsx \
  -cut_freq 0.75 \
  -cut_lin 10 \
  -g \
  -out output/matrix/desh_2023-02-01--2023-04-01
```
![Mutation Frequency Matrix Example](src/example_matrix.png)
*Figure 1: Example mutation frequency matrix illustrating lineage-specific mutation patterns.*

Output files include a frequency matrix and a meta data file with name of the parent lineage in pango annotation, the frequency of the lineage within the time frame and in how many labs the lineages were sequenced.

### variant-specific pcr test-design
signature mutations can be used to determine which mutations acuretly define a lineage (nothing else) which can be used to test for specific lineage in a variant-specific pcr test-design.

How signature mutations can be calculated:  

```bash
python3 scripts/init.py -tsv input/covsonar-rki-2023-02-01--2023-04-01.tsv -sig
```

### consensus^2 for representing a lineage

How a consensus sequence from lineages within the time frame can be reconstructed:  

```bash
python3 scripts/init.py -tsv input/covsonar-rki-2023-02-01--2023-04-01.tsv -con
```

---

## Cite SC2MFC 
If you use SC2MFC in scientific publications, please cite:
```
Ghassemi, A. (2023). SC2 Mutation Frequency Calculator (SC2MFC). Zenodo.
https://doi.org/10.5281/zenodo.17807860
```

---

## Contribution 
You can contribute to this project by reporting problems or writing feature requests to the issue section under https://github.com/rki-mf1/sc2-mutation-frequency-calculator/issues

Your feedback is very welcome!

---
## Contact
Maintainer: Ashkan Ghassemi (ashkan98)




